hydromt/models/model_mesh.py

```diff
@@ -253,33 +293,163 @@ def set_mesh(
                     f"Cannot set mesh from {str(type(data).__name__)} without a name."
                 )
             data = data.to_dataset()
+
+        # Checks on grid topology
+        # TODO: check if we support setting multiple grids at once. For now just one
+        if len(data.ugrid.grids) > 1:
+            raise ValueError(
+                "set_mesh methods only supports adding data to one grid at a time."
+            )
+        if grid_name is None:
+            grid_name = data.ugrid.grid.name
+        elif grid_name != data.ugrid.grid.name:
+            data = workflows.rename_mesh(data, name=grid_name)
+
+        # Adding to mesh
         if self._mesh is None:  # NOTE: mesh is initialized with None
+            # Check on crs
+            if not data.ugrid.grid.crs:
+                raise ValueError("Data should have CRS.")
             self._mesh = data
         else:
-            for dvar in data.data_vars:
-                if dvar in self._mesh:
-                    self.logger.warning(f"Replacing mesh parameter: {dvar}")
-                self._mesh[dvar] = data[dvar]
+            # Check on crs
+            if not data.ugrid.grid.crs == self.crs:
+                raise ValueError("Data and self.mesh should have the same CRS.")
+            # Save crs as it will be lost when converting to xarray
+            crs = self.crs
+            # Check on new grid topology
+            if grid_name in self.mesh_names:
+                # check if the two grids are the same
+                if (
+                    not self.mesh_grids[grid_name]
+                    .to_dataset()
+                    .equals(data.ugrid.grid.to_dataset())
+                ):
+                    if not overwrite_grid:
+                        raise ValueError(
+                            f"Grid {grid_name} already exists in mesh"
+                            " and has a different topology. "
+                            "Use overwrite_grid=True to overwrite the grid"
+                            " topology and related data."
+                        )
+                    else:
+                        # Remove grid and all corresponding data variables from mesh
+                        self.logger.warning(
+                            f"Overwriting grid {grid_name} and the corresponding"
+                            " data variables in mesh."
+                        )
+                        grids = [
+                            self.mesh_datasets[g].ugrid.to_dataset()
+                            for g in self.mesh_names
+                            if g != grid_name
+                        ]
+                        # Re-define _mesh
+                        grids = xr.merge(grids)
+                        self._mesh = xu.UgridDataset(grids)
```

Comment:
Should also preserve optional attributes
```suggestion
                            self.mesh_datasets[g].ugrid.to_dataset(optional_attributes=True)
```

Reply:
Same as above